### ISSUE TITLE
前向Op的计算结果的内存如何可写的共享给反向Op

### ISSUE THREAD

**Issue by lcy-seso** [PaddlePaddle/Paddle#4082]
在实现 `softmax_with_cross_entropy_op` 时遇到一个问题，简单来说，`softmax_with_cross_entropy_op` 前向和反向计算过程如下：

- 前向：输入 `X`和`groundTruth`，
    1. 对输入`X` 计算 `softmax`，得到概率输出`softmax_out`；
    2.  以 `softmax` 和 `grountTruth` 计算 `cross_entropy`

- 反向：回传给 `X` 的梯度是：`softmax_out` - `groundTruth` （这里省略掉维度相关的计算细节）

`softmax` 和 `cross_entropy` 将两个 OP 组合为一个 OP的优点有以下两点：

1.  消掉中间步骤的一些指数项，提高数值计算的稳定性；
2. 只计算一次 softmax，反向也是softmax再做一次减法。

道理上，**如果前向直接将`softmax_out` 的计算结果存储在回传给 X 的梯度中，反向再做一次减法，这样可以分配更少的内存。**

在现在的设计中，前向Op的Output会成为反向的Input，反向可以通过这样的方式得到前向计算的一些中间结果再次使用。

在`ExecutionContext` 中拿到的Input 是 const ，无法再修改其中的内容。

在反向 Op 中得到了前向Softmax 计算结果，依然需要拷贝至输入 X 的梯度矩阵，然后才可以做减法，（Tensor 中有 `ShareDataWith` 接口）。

请问：
1. 是否有什么方式可以将前向计算的中间结果，以可写的方式 share 给反向来减少这一份内存的拷贝。
2. 或者这种需要违背了框架的整体设计，牺牲内存空间来维护设计的统一性？
3. 或者目前是可以支持这样的需要，是我理解不对呢？

**Comment by lcy-seso** [comment]
Tensor 的 ShareDataWith 接口应该是可以。
